### ISSUE TITLE
Incorrect CS8762 for null coalescing return

### ISSUE THREAD

**Issue by sharwell** [dotnet/roslyn#43241]
**Version Used**: 16.6 Preview 2

**Steps to Reproduce**:

```csharp
public bool TryGet(int key, [NotNullWhen(true)] out TriviaData? triviaInfo)
{
    triviaInfo = null;
    return _map?.TryGetValue(key, out triviaInfo) ?? false;
}
```

**Expected Behavior**:

No warning.

**Actual Behavior**:

CS8762: Parameter 'triviaInfo' may not have a null value when exiting with 'true'.

If the code is converted to the equivalent "full" code with explicit null checks, no warning is reported.

**Comment by RikkiGibson** [comment]
I suspect that there is an issue analyzing conditional-access calls with 'out' arguments.. we just aren't able to deduce that when this expression returns 'true', then we definitely assigned something to 'triviaInfo'.

a few questions:
1. what is the signature of the `TryGetValue` method being invoked?
2. could you specify what equivalent full code you used? I imagine `_map != null && _map.TryGetValue(key, out triviaInfo)` might not report the warning.

**Comment by sharwell** [comment]
This was the full form:

```csharp
triviaInfo = null;
if (_map is object)
  return _map.TryGetValue(key, out triviaInfo);

return false;
```

The `_map` is `ConcurrentDictionary<int, TriviaData>`.

**Comment by FiniteReality** [comment]
I'm not sure if this is the same issue, but this also seems to occur on ternary operators:
```cs
public static bool TryParseAddress(ReadOnlySequence<byte> sequence,
    [NotNullWhen(true)] out IPAddress? address)
{
    return !sequence.IsSingleSegment
        ? SlowPath(sequence, out address)
        : TryParseAddress(sequence.FirstSpan, out address);
}
```

Both `SlowPath` and `TryParseAddress(ReadOnlySpan<byte>, out IPAddress? address)` have the same annotations as the method shown, and all code paths lead to `IPAddress.TryParse`, whose out parameter itself is marked with `NotNullWhen(true)`.

**Comment by jcouv** [comment]
I think this was fixed. We've relaxed the enforcement of NotNullWhen attributes on method bodies (maybe in 16.8?).
Now we only warn when the nullability states differ between the WhenTrue and WhenFalse branches at the return point. Typically, this means having returned a constant (true/false) or another method with attributes.

Please provide a self-contained repro if you think the issue was not satisfactorily resolved. Thanks
